docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,378 @@
+  string trace = 1;
+  // base denomination of the relayed fungible token
+  string base_denom = 2;
+}
+```
+
+The `IBCDenom` function constructs the `Coin` denomination used when creating the ICS20 fungible token packet data:
+
+```golang
+// Hash returns the hex bytes of the SHA256 hash of the DenomTrace fields.
+func (dt DenomTrace) Hash() tmbytes.HexBytes {
+  return tmhash.Sum(dt.Trace + "/" + dt.BaseDenom)
+}
+
+// IBCDenom a coin denomination for an ICS20 fungible token in the format 'ibc/{hash(trace + baseDenom)}'. If the trace is empty, it will return the base denomination.
+func (dt DenomTrace) IBCDenom() string {
+  if dt.Trace != "" {
+    return fmt.Sprintf("ibc/%s", dt.Hash())
+  }
+  return dt.BaseDenom
+}
+```
+
+In order to trim the denomination trace prefix when sending/receiving fungible tokens, the `RemovePrefix` funciton is provided.
+
+> NOTE: the prefix addition must be done on the client side.
+
+```golang
+// RemovePrefix trims the first portID/channelID pair from the trace info. If the trace is already empty it will perform a no-op. If the trace is incorrectly constructed or doesn't have separators it will return an error.
+func (dt *DenomTrace) RemovePrefix() error {
+  if dt.Trace == "" {
+    return nil
+  }
+
+  traceSplit := strings.SplitN(dt.Trace, "/", 3)
+
+  var err error
+  switch {
+  case len(traceSplit) == 0, traceSplit[0] == dt.Trace:
+    err = Wrapf(ErrInvalidDenomForTransfer, "trace info %s must contain '/' separators", dt.Trace)
+  case len(traceSplit) == 1:
+    err = Wrapf(ErrInvalidDenomForTransfer, "trace info %s must come in pairs of '{portID}/channelID}'", dt.Trace)
+  case len(traceSplit) == 2:
+    dt.Trace = ""
+  case len(traceSplit) == 3:
+    dt.Trace = traceSplit[2]
+  }
+
+  if err != nil {
+    return err
+  }
+
+  return nil
+}
+```
+
+### `x/ibc-transfer` Changes
+
+In order to retreive the trace information from an IBC denomination, a lookup table needs to be
+added to the `ibc-transfer` module. These values need to also be persisted between upgrades, meaning
+that a new `[]DenomTrace` `GenesisState` field state needs to be added to the module:
+
+```golang
+// GetDenom retreives the full identifiers trace and base denomination from the store.
+func (k Keeper) GetDenomTrace(ctx Context, denomTraceHash []byte) (DenomTrace, bool) {
+  store := ctx.KVStore(k.storeKey)
+  bz := store.Get(types.KeyDenomTrace(traceHash))
+  if bz == nil {
+    return &DenomTrace, false
+  }
+
+  var denomTrace DenomTrace
+  k.cdc.MustUnmarshalBinaryBare(bz, &denomTrace)
+  return denomTrace, true
+}
+
+// HasTrace checks if a the key with the given trace hash exists on the store.
+func (k Keeper) HasDenomTrace(ctx Context, denomTraceHash []byte)  bool {
+  store := ctx.KVStore(k.storeKey)
+  return store.Has(types.KeyTrace(denomTraceHash))
+}
+
+// SetTrace sets a new {trace hash -> trace} pair to the store.
+func (k Keeper) SetTrace(ctx Context, denomTraceHash []byte, denomTrace DenomTrace) {
```

Comment:
Only the `denomTrace` parameter is needed because the hash is derived from it and should be verified anyway.